packages/gatsby-source-contentful/src/gatsby-node.js

```diff
@@ -112,6 +119,121 @@ exports.sourceNodes = async (
     pluginConfig,
   })
 
+  createTypes(`
+  interface ContentfulEntry @nodeInterface {
+    contentful_id: String!
+    id: ID!
+  }
+`)
+
+  createTypes(`
+  interface ContentfulReference {
+    contentful_id: String!
+    id: ID!
+  }
+`)
+
+  createTypes(
+    schema.buildObjectType({
+      name: `ContentfulAsset`,
+      fields: {
+        contentful_id: { type: `String!` },
+        id: { type: `ID!` },
+      },
+      interfaces: [`ContentfulReference`, `Node`],
+    })
+  )
+
+  const gqlTypes = contentTypeItems.map(contentTypeItem =>
+    schema.buildObjectType({
+      name: _.upperFirst(
+        _.camelCase(
+          `Contentful ${
+            pluginConfig.get(`useNameForId`)
+              ? contentTypeItem.name
+              : contentTypeItem.sys.id
+          }`
+        )
+      ),
+      fields: {
+        contentful_id: { type: `String!` },
+        id: { type: `ID!` },
+      },
+      interfaces: [`ContentfulReference`, `ContentfulEntry`, `Node`],
+    })
+  )
+  createTypes(gqlTypes)
+
+  console.time(`Process Contentful data`)
+
+  // Remove deleted entries and assets from cached sync data set
+  previousSyncData.entries = previousSyncData.entries.filter(
+    entry =>
+      _.findIndex(
+        currentSyncData.deletedEntries,
+        o => o.sys.id === entry.sys.id
+      ) !== -1
+  )
+
+  previousSyncData.assets = previousSyncData.assets.filter(
+    asset =>
+      _.findIndex(
+        currentSyncData.deletedAssets,
+        o => o.sys.id === asset.sys.id
+      ) !== -1
+  )
+
+  // Merge cached data with new sync data
+  // order is important here, fresh data first
+  currentSyncData.entries = _.uniqBy(
+    currentSyncData.entries.concat(previousSyncData.entries),
+    `sys.id`
+  )
+
+  currentSyncData.assets = _.uniqBy(
+    currentSyncData.assets.concat(previousSyncData.assets),
+    `sys.id`
+  )
+
+  // Store a raw and unresolved copy of the data for caching
+  const currentSyncDataRaw = _.cloneDeep(currentSyncData)
+
+  // Use the JS-SDK to resolve the entries and assets
+  const res = client.parseEntries({
+    items: currentSyncData.entries,
+    includes: {
+      assets: currentSyncData.assets,
+      entries: currentSyncData.entries,
+    },
+  })
+
+  currentSyncData.entries = res.items
+
+  // Inject raw API output to rich text fields
+  const richTextFields = contentTypeItems.reduce((fields, contentType) => {
+    return {
+      ...fields,
+      [contentType.sys.id]: contentType.fields
+        .filter(field => field.type === `RichText`)
+        .map(field => field.id),
+    }
+  }, {})
```

Comment:
Reduce is almost never the answer here
```suggestion
  const richTextFields = {};
  contentTypeItems.forEach((contentType) => 
    richTextFields[contentType.sys.id] = contentType.fields
        .filter(field => field.type === `RichText`)
        .map(field => field.id),
  );
```

I think `richTextFields` should be a `Map` here, btw. The only use is on the next block where a `.get` should be faster than using an object as a hash.